x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/InlineStats.java

```diff
@@ -77,4 +135,63 @@ public boolean equals(Object obj) {
             && Objects.equals(aggregates, other.aggregates)
             && Objects.equals(child(), other.child());
     }
+
+    @Override
+    public LogicalPlan firstPhase() {
+        return new Aggregate(source(), child(), Aggregate.AggregateType.STANDARD, groupings, aggregates);
+    }
+
+    @Override
+    public LogicalPlan nextPhase(List<Attribute> schema, List<Page> firstPhaseResult) {
+        LocalRelation local = firstPhaseResultsToLocalRelation(schema, firstPhaseResult);
+        List<Attribute> groupingAttributes = new ArrayList<>(groupings.size());
+        for (Expression g : groupings) {
+            if (g instanceof Attribute a) {
+                groupingAttributes.add(a);
+            } else {
+                throw new UnsupportedOperationException("INLINESTATS doesn't support expressions in grouping position yet");
+            }
+        }
+        List<Attribute> leftFields = new ArrayList<>(groupingAttributes.size());
+        List<Attribute> rightFields = new ArrayList<>(groupingAttributes.size());
+        List<Attribute> rhsOutput = Join.makeReference(local.output());
+        for (Attribute lhs : groupingAttributes) {
+            for (Attribute rhs : rhsOutput) {
+                if (lhs.name().equals(rhs.name())) {
+                    leftFields.add(lhs);
+                    rightFields.add(rhs);
+                    break;
+                }
+            }
+        }
+        JoinConfig config = new JoinConfig(JoinType.LEFT, groupingAttributes, leftFields, rightFields);
+        return new Join(source(), child(), local, config);
+    }
+
+    private LocalRelation firstPhaseResultsToLocalRelation(List<Attribute> schema, List<Page> firstPhaseResult) {
+        // Limit ourselves to 1mb of results similar to LOOKUP for now.
+        long bytesUsed = firstPhaseResult.stream().mapToLong(Page::ramBytesUsedByBlocks).sum();
+        if (bytesUsed > ByteSizeValue.ofMb(1).getBytes()) {
+            throw new IllegalArgumentException("first phase result too large [" + ByteSizeValue.ofBytes(bytesUsed) + "] > 1mb");
```

Comment:
Isn't it a bit late to do it here? As I understand it, we already computed everything and got the data into the coordinator.
I guess it's better to have a limit than not have anything; just to understand it

Reply:
In an ideal world we wouldn't have a limit here but we don't have proper accounting for `Block`s in the request objects - they don't use the Breaker at all. So doing this copy shifts it out of tracking-land. I think it's important to get the tracking right before we GA this so I'm adding that to the follow-ups list. I want to keep it out of this PR because I don't really know how to do it and it's going to touch a bunch of stuff.